### PR TITLE
added hats finance v2 support

### PR DESCRIPTION
**Pull request by shayzluf**
**NOTE**

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. The protocol is usually listed within 24 hours of merging the PR
4. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
5. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
6. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
7. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Github org/user (Optional, if your code is open source, we can track activity):

### CODE REVIEW

**Comment by llamatester** [comment]
The adapter at projects/hats exports TVL: 
        
 
 
    polygon                   0
    optimism                  0
    ethereum                  2.08 M
    arbitrum                  45.80 k
    
    total                    2.13 M

File: projects/hats/index.js [hunk 1]
```diff
@@ -1,5 +1,122 @@
...
+const chain = 'ethereum'; // The chain the MasterChef contract is on
+
+const masterChefTVL = masterChefExports(owner, chain, token);
+
+async function tvlCheff(timestamp, block) {
+    let balances = {};
+    // Get the TVL from the MasterChef contract
+    const chainBlocks = { ethereum: block };
+    // Add the MasterChef TVL to the balances
+    const masterChefBalances = await masterChefTVL[chain].tvl(timestamp, block, chainBlocks);
```
Comment (g1nt0ki) [inline projects/hats/index.js:28]: v1 vaults are included in graph leading to doublecounting here

**Comment by g1nt0ki** [comment]
hi @shayzluf , thanks for the PR.

I try to avoid graph where I can, so I made an implementation here with on chain calls, the numbers match with the ones from the graph, also I think it is better to have a new listing for v2: https://github.com/DefiLlama/DefiLlama-Adapters/commit/270980f3f51e123f5f639d95c5d3b72b0f2b9764

Let me know if I have introduced any bugs.

**Comment by shayzluf** [comment]
> hi @shayzluf , thanks for the PR.
> 
> I try to avoid graph where I can, so I made an implementation here with on chain calls, the numbers match with the ones from the graph, also I think it is better to have a new listing for v2: [270980f](https://github.com/DefiLlama/DefiLlama-Adapters/commit/270980f3f51e123f5f639d95c5d3b72b0f2b9764)
> 
> Let me know if I have introduced any bugs.

Hats v2 is not a mastercheff clone and it spawns up other contracts (HatVault) fro each bug bounty.  We could minimally use the graph to get all the vault lists and then just get their balance from the provider . or get all the vaultcreated events and use just the provider. WDYT?

**Comment by g1nt0ki** [comment]
hi @shayzluf, yeah I agree hats v2 is not masterchef. 

"or get all the vaultcreated events and use just the provider. " this is what my code is doing right?

**Comment by g1nt0ki** [comment]
ah, sorry realized not using logs here, but all the created vaults can be found with `hatVaults(uint256)`  method right?

**Comment by shayzluf** [comment]
> ah, sorry realized not using logs here, but all the created vaults can be found with `hatVaults(uint256)`  method right?

I think, so

Checking

**Pull request closed without merging**